wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs

```diff
@@ -1,43 +1,233 @@
 bracketPassiveWallet
-    :: forall ctx m n a. (MonadMask m, WalletDbReader ctx n, MonadIO n, MonadThrow n)
-    => (PassiveWalletLayer n -> m a) -> m a
+    :: forall ctx m n a. (MonadMask n, MonadLegacyWallet ctx m)
+    => (PassiveWalletLayer m -> n a) -> n a
 bracketPassiveWallet =
     bracket
         (pure passiveWalletLayer)
         (\_ -> return ())
   where
-    passiveWalletLayer :: PassiveWalletLayer n
+    passiveWalletLayer :: PassiveWalletLayer m
     passiveWalletLayer = PassiveWalletLayer
-        { pwlGetWalletIds  = askWalletSnapshot >>= \ws -> migrate $ getWalletAddresses ws
+        { _pwlCreateWallet  = pwlCreateWallet
+        , _pwlGetWalletIds  = pwlGetWalletIds
+        , _pwlGetWallet     = pwlGetWallet
+        , _pwlUpdateWallet  = pwlUpdateWallet
+        , _pwlDeleteWallet  = pwlDeleteWallet
+
+        , _pwlCreateAccount = pwlCreateAccount
+        , _pwlGetAccounts   = pwlGetAccounts
+        , _pwlGetAccount    = pwlGetAccount
+        , _pwlUpdateAccount = pwlUpdateAccount
+        , _pwlDeleteAccount = pwlDeleteAccount
+
+        , _pwlGetAddresses  = pwlGetAddresses
         }
 
+
 -- | Initialize the active wallet.
 -- The active wallet is allowed all.
 bracketActiveWallet
-    :: forall ctx m n a. (MonadMask m, WalletDbReader ctx n, MonadIO n, MonadThrow n)
-    => PassiveWalletLayer n
+    :: forall ctx m n a. (MonadMask n, MonadLegacyWallet ctx m)
+    => PassiveWalletLayer m
     -> WalletDiffusion
-    -> (ActiveWalletLayer n -> m a) -> m a
+    -> (ActiveWalletLayer m -> n a) -> n a
 bracketActiveWallet walletPassiveLayer walletDiffusion =
     bracket
       (return ActiveWalletLayer{..})
       (\_ -> return ())
 
+------------------------------------------------------------
+-- Wallet
+------------------------------------------------------------
+
+pwlCreateWallet
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => NewWallet
+    -> m Wallet
+pwlCreateWallet NewWallet{..} = do
+
+    let spendingPassword = fromMaybe mempty $ coerce newwalSpendingPassword
+    let backupPhrase     = coerce newwalBackupPhrase
+
+    initMeta    <- CWalletMeta  <$> pure newwalName
+                                <*> migrate newwalAssuranceLevel
+                                <*> pure 0
+
+    let walletInit = CWalletInit initMeta backupPhrase
+
+    wallet      <- newWalletHandler newwalOperation spendingPassword walletInit
+    wId         <- migrate $ cwId wallet
+
+    -- Get wallet or throw if missing.
+    maybeThrow (MissingWallet wId) =<< pwlGetWallet wId
+  where
+    -- | We have two functions which are very similar.
+    newWalletHandler :: WalletOperation -> PassPhrase -> CWalletInit -> m CWallet
+    newWalletHandler CreateWallet  = newWallet
+    newWalletHandler RestoreWallet = restoreWallet
+
+
+pwlGetWalletIds
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => m [WalletId]
+pwlGetWalletIds = do
+    ws          <- askWalletSnapshot
+    migrate $ getWalletAddresses ws
```

Comment:
Perhaps it's me but the naming really confused me here. I know it is from the legacy wallet but this apparently returns wallet ids. At least the new passive layer fixes the naming :)